Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java

```diff
@@ -258,6 +278,57 @@ private static void checkAssertion(SamlAssertionWrapper assertion, String confir
             }
         }
 
+        if (validateAttributes()) {
+            checkAttributes(assertion);
+        }
+    }
+
+    public void checkAttributes(SamlAssertionWrapper assertion) throws WSSecurityException {
+        Assertion saml2Assertion = assertion.getSaml2();
+        Set<String> foundAttr = new HashSet<>();
+
+        for (AttributeStatement statement : saml2Assertion.getAttributeStatements()) {
+            for (Attribute attribute : statement.getAttributes()) {
+                if (CollectionUtils.isNotEmpty(attribute.getAttributeValues()) && attribute.getAttributeValues().get(0) != null
+                        && isAttributeValueValid(attribute.getAttributeValues().get(0), attribute.getName())) {
+                    foundAttr.add(attribute.getName());
+                }
+            }
+        }
+
+        Set<String> missingAttrs = new HashSet<>(NhincConstants.VALIDATED_ATTRIBUTES);
+        missingAttrs.removeAll(foundAttr);
+
+        if (CollectionUtils.isNotEmpty(missingAttrs)) {
+            LOG.error("Missing Required Attributes: {}", missingAttrs);
+            throw new WSSecurityException(ErrorCode.FAILED_CHECK, "missingRequiredSAMLAttribute");
+        }
+    }
+
+    private boolean isAttributeValueValid(XMLObject value, String name) {
+        if (value instanceof XSStringImpl) {
+            return NullChecker.isNotNullish(((XSStringImpl) value).getValue());
+        } else if (value instanceof XSAnyImpl) {
+            if (CollectionUtils.isNotEmpty(((XSAnyImpl) value).getUnknownXMLObjects())
+                    && ((XSAnyImpl) value).getUnknownXMLObjects().get(0) instanceof XSAnyImpl) {
+                XSAnyImpl attributeValues = (XSAnyImpl) ((XSAnyImpl) value).getUnknownXMLObjects().get(0);
+                if (!attributeValues.getUnknownAttributes().isEmpty()) {
+                    return containsCodeQName(attributeValues.getUnknownAttributes());
+                } 
+            }
+            return false;
+        }
+        LOG.warn("Unknown type of Attribute object for name: {}, default to add to found list.", name);
+        return true;
+    }
+    
+    private boolean containsCodeQName(AttributeMap unknownAttributes) {
+        for(QName qName : unknownAttributes.keySet()) {
+            if(qName.getLocalPart().equals(NhincConstants.CE_CODE_ID)) {
```

Comment:
Should we also need to check for codeSystem as well since both always go together?
Optional: Consider to compare using constant.equalsIgnoreCase(variable) approach to prevent NPE?

Reply:
I'll make change for NPE, but don't need check of codesystem.  Overkill and codesystem is a constant.